File: TextClassifier/classifier.ml
open Core
open ExtList
open Lacaml.D
open Libsvm

module StringSet = Set.Make(String)
module SM = Map.Make(String)

module type Classifier = sig
  val train : string DynArray.t SM.t -> unit
  val classify : string -> string
end

let (|>) x f = f x

let flip f x y = f y x

let find_default m k def =
  try
    SM.find k m
  with Not_found ->
    def

let process_file filename f =
  let chan = open_in filename in
  try
    while true; do
      input_line chan |> f
    done;
  with End_of_file ->
    close_in chan

let read_words filename f =
  let islower c = let c = Char.chr c in 'a' <= c && c <= 'z' in
  let chan = open_in filename in
  let read_char () =
    try
      input_char chan |> Char.lowercase |> Char.code
    with End_of_file -> (
      close_in chan;
      -1
    ) in
  let rec go flag s = match flag with
    | false ->
        let c = read_char () in
        if c = -1 then
          ()
        else if islower c then
          go true (String.make 1 (Char.chr c))
        else
          go false s
    | true ->
        let c = read_char () in
        if c = -1 then
          f s
        else if islower c then
          go true (s ^ String.make 1 (Char.chr c))
        else (
          f s;
          go false s
        ) in
  go false ""

module Idf = struct
  let ndoc = ref 0
  let vocab = ref SM.empty
  let idf = ref SM.empty
  let categories = DynArray.create ()
  let cat2id = ref SM.empty

  let calc_idf cat2filenames =
    let cat_id = ref 0 in
    let vocab_id = ref 0 in
    ndoc := SM.fold (fun _ filenames sum ->
      sum + DynArray.length filenames) cat2filenames 0;
    SM.iter (fun cat filenames ->
      DynArray.add categories cat;
      cat2id := SM.add cat !cat_id !cat2id;
      incr cat_id;
      let added = ref StringSet.empty in
      DynArray.iter (fun filename ->
        read_words filename (fun word ->
          if SM.mem word !vocab |> not then (
            vocab := SM.add word !vocab_id !vocab;
            incr vocab_id
          );

          (* document frequency *)
          if StringSet.mem word !added |> not then (
            added := StringSet.add word !added;
            let dfv = find_default !idf word 0.0 in
            idf := SM.add word (dfv +. 1.0) !idf
          );
        );
      ) filenames;
    ) cat2filenames;
    (* inverse document frequency *)
    idf := SM.map (fun dfv -> log (float_of_int !ndoc /. dfv)) !idf;
    print_endline "meow"
end

module NaiveBayes = struct
  include Idf

  let prior = ref SM.empty

  let prob = ref SM.empty

  let train cat2filenames =
    let ncat = SM.fold (fun _ _ acc -> acc + 1) cat2filenames 0 in

    calc_idf cat2filenames;

    prior := SM.map (fun filenames ->
      (float_of_int (DynArray.length filenames) /.  float_of_int ncat))
      cat2filenames;

    SM.iter (fun cat filenames ->
      let cat_prob = ref SM.empty in
      DynArray.iter (fun filename ->
        print_endline filename;
        read_words filename (fun word ->
          let cnt = find_default !cat_prob word 0.0 in
          cat_prob := SM.add word (cnt +. 1.0) !cat_prob;
        );
        print_endline filename;
      ) filenames;
      let tot = SM.fold (fun _ c acc -> acc +. c +. 1.0) !cat_prob 0.0 in
      SM.iter (fun word _ ->
        let cnt = find_default !cat_prob word 0.0 in
        cat_prob := SM.add word ((cnt +. 1.0) /. tot) !cat_prob
      ) !vocab;
      prob := SM.add cat !cat_prob !prob;
    ) cat2filenames

  let classify filename =
    let score = prior in
    read_words filename (fun word ->
      if SM.mem word !vocab then
        SM.iter (fun cat cat_prob ->
          let value = SM.find cat !score in
          let p = SM.find word cat_prob in
          score := SM.add cat (value +. log p) !score
        ) !prob
    );
    SM.fold (fun cat v ((optc, optv) as opt) ->
      if v > optv then (cat, v) else opt
    ) !score ("", neg_infinity) |> fst
end

module SVM = struct
  include Idf

  open Lacaml.D

  let model =
    let x = Mat.of_array
        [|
          [| 0.; 0. |];
          [| 0.; 1. |];
          [| 1.; 0. |];
          [| 1.; 1. |];
        |]
      in
    let targets = Vec.of_array [| 0.; 1.; 1.; 0. |] in
    let problem = Svm.Problem.create ~x ~y:targets in
    ref (Svm.train ~kernel:`RBF problem)

  let features = ref SM.empty

  let train cat2filenames =
    calc_idf cat2filenames;

    (*Printf.printf "%d \n" !ndoc ;*)
    SM.fold (fun k v acc -> (k, v) :: acc) !idf [] |>
      List.sort ~cmp:compare |> List.take 20000 |>
      List.iteri (fun i (word, _) -> features := SM.add word i !features);

    let nfeatures = SM.fold (fun _ _ acc -> acc + 1) !features 0 in
    (*Printf.printf "%d %d\n" !ndoc (Hashtbl.length features);*)

    let x = Mat.make !ndoc nfeatures 0.0 in
    let y = Vec.create !ndoc in
    let row = ref 0 in

    SM.iter (fun cat filenames ->
      let cat_id = SM.find cat !cat2id in
      DynArray.iter (fun filename ->
        incr row;
        read_words filename (fun word ->
          if SM.mem word !features then (
            let id = SM.find word !features + 1 in
            let idfv = SM.find word !idf in
            x.{!row, id} <- x.{!row, id} +. idfv
          )
        );
        y.{!row} <- float_of_int cat_id
      ) filenames
    ) cat2filenames;

    let problem = Svm.Problem.create ~x ~y in
    model := Svm.train ~kernel:`RBF problem

  let classify filename =
    let nfeatures = SM.fold (fun _ _ acc -> acc + 1) !features 0 in
    let x = Mat.make 1 nfeatures 0.0 in
    read_words filename (fun word ->
      if SM.mem word !features then (
        let id = SM.find word !features + 1 in
        let idfv = SM.find word !idf in
        x.{1, id} <- x.{1, id} +. idfv
      )
    );
    let y = Float.round (Svm.predict !model ~x).{1} |> int_of_float
      |> max 0 |> min (DynArray.length categories) in
    DynArray.get categories y
end

module MakeClassifier (C: Classifier) = struct
  let process_filelist filelist cont =
    let cat = ref "" in
    process_file filelist (fun line ->
      try
        Scanf.sscanf line "# %s" (fun cat2 -> cat := cat2; cont !cat "")
      with Scanf.Scan_failure e ->
        cont !cat line
      )

  let train filelist =
    let cat2filenames = ref SM.empty in
    process_filelist filelist (fun cat filename ->
      if filename = "" then
        cat2filenames := SM.add cat (DynArray.create ()) !cat2filenames
      else
        let arr = SM.find cat !cat2filenames in
        DynArray.add arr filename
    );
    C.train !cat2filenames

  let classify filelist =
    let correct = ref 0 in
    let tot = ref 0 in
    process_filelist filelist (fun cat filename ->
      if filename = "" then
        Printf.printf "# %s\n" cat
      else (
        tot := !tot + 1;
        if C.classify filename = cat then
          correct := !correct + 1
      )
    );
    Printf.printf "precision: %f\n" (float_of_int !correct /. float_of_int !tot)
end

module NBClassifier = MakeClassifier(NaiveBayes);;

NBClassifier.train "train.list";;

NBClassifier.classify "classify.list";;

(*module SVMClassifier = MakeClassifier(SVM);;*)

(*SVMClassifier.train "train.list";;*)

(*SVMClassifier.classify "classify.list";;*)

File: TextClassifier/classifier.cc
#include <algorithm>
#include <cmath>
#include <fstream>
#include <functional>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>
#include <svm.h>
using namespace std;

class Word
{
public:
  Word(string filename) : fin(filename) {
  }
  void operator()(function<void(string)> f) {
    string buf;
    bool flag = false;
    for(;;) {
      char c = tolower(fin.get());
      if (flag) {
        if (! fin) {
          f(buf);
          break;
        } else if (islower(c))
          buf += c;
        else {
          f(buf);
          flag = false;
        }
      } else {
        if (! fin)
          break;
        else if (islower(c)) {
          flag = true;
          buf = c;
        }
      }
    }
  }
protected:
  ifstream fin;
};

template<typename K, typename V>
V findDefault(const map<K,V> &m, K k, V def = V())
{
  return m.count(k) ? m[k] : def;
}

class Classifier
{
public:
  virtual void train(map<string, vector<string>> cat2f) = 0;
  virtual string classify(string) = 0;
protected:
  void calcIdf(map<string, vector<string>> cat2f) {
    ndoc = 0;

    for (auto i: cat2f) {
      string cat = i.first;
      ndoc += i.second.size();
      categories.push_back(cat);
      for (auto filename: i.second) {
        set<string> added;
        Word reader(filename);
        reader([&](string w) {
          if (! vocab.count(w)) {
            int t = int(vocab.size());
            vocab[w] = t;
          }
          // document frequency
          if (! added.count(w)) {
            added.insert(w);
            idf[vocab[w]] += 1.0;
          }
        });
      }
    }

    // df -> idf
    for (auto &i: idf)
      i.second = log((ndoc + 1.0) / i.second);
  }

  int ndoc;
  map<string, int> vocab;
  map<int, double> idf;
  vector<string> categories;
};

class NaiveBayes : public Classifier
{
public:
  virtual void train(map<string, vector<string>> cat2f) override {
    calcIdf(cat2f);
    prior.resize(categories.size());
    prob.resize(categories.size());

    for (size_t i = 0; i < categories.size(); i++) {
      string cat = categories[i];
      prior[i] = double(cat2f[cat].size()) / ndoc;

      for (auto filename: cat2f[cat]) {
        Word reader(filename);
        reader([&](string w) {
          prob[i][vocab[w]] += 1.0;
        });
      }

      double tot = 0.0;
      for (auto w: prob[i])
        tot += w.second + 1.0;
      for (auto w: vocab)
        prob[i][w.second] = (prob[i][w.second] + 1.0) / tot;
    }
  }

  virtual string classify(string filename) override {
    vector<double> score(prior);
    Word reader(filename);
    reader([&](string w) {
      if (vocab.count(w))
        for (size_t i = 0; i < categories.size(); i++)
          score[i] += log(prob[i][vocab[w]]);
    });
    return categories[max_element(score.begin(), score.end()) - score.begin()];
  }

protected:
  vector<double> prior;
  vector<map<int, double> > prob;
};

class SVM : public Classifier
{
public:
  ~SVM() {
    svm_free_model_content(model);

    for (int i = 0; i < problem.l; i++)
      delete[] problem.x[i];
    delete[] problem.x;
    delete[] problem.y;
  }

  virtual void train(map<string, vector<string>> cat2f) override {
    calcIdf(cat2f);

    problem.l = ndoc;
    problem.y = new double[ndoc];
    problem.x = new svm_node*[ndoc];

    int row = 0;
    for (size_t i = 0; i < categories.size(); i++) {
      string cat = categories[i];

      for (auto filename: cat2f[cat]) {
        map<int, int> ws;
        Word reader(filename);
        reader([&](string w) {
          if (vocab.count(w))
            ws[vocab[w]]++;
        });

        int col = 0;
        problem.y[row] = i;
        problem.x[row] = new svm_node[ws.size() + 1];
        for (auto j: ws) {
          problem.x[row][col].index = j.first;
          problem.x[row][col].value = j.second * idf[j.first]; // tf*idf
          col++;
        }
        problem.x[row][col].index = -1;
        normalize(col, problem.x[row]);
        row++;
      }
    }

    svm_parameter param;
    param.svm_type = C_SVC;
    param.kernel_type = LINEAR;
    param.cache_size = 100;
    param.eps = 1e-5;
    param.C = 1;
    param.nr_weight = 0;
    param.weight_label = NULL;
    param.weight = NULL;
    param.shrinking = 1;
    param.probability = 0;

    const char *pstr = svm_check_parameter(&problem, &param);
    if (pstr)
      puts(pstr);
    model = svm_train(&problem, &param);
  }

  virtual string classify(string filename) override {
    map<int, int> ws;
    Word reader(filename);
    reader([&](string w) {
      if (vocab.count(w))
        ws[vocab[w]]++;
    });

    int col = 0;
    svm_node *x = new svm_node[ws.size() + 1];
    for (auto j: ws) {
      x[col].index = j.first;
      x[col].value = j.second * idf[j.first]; // tf*idf
      col++;
    }
    x[col].index = -1;
    normalize(col, x);

    double res = svm_predict(model, x);
    delete[] x;
    return categories[int(res)];
  }

protected:
  void normalize(int n, svm_node x[]) {
    double scale = 0;
    for (int i = 0; i < n; i++)
      scale += x[i].value * x[i].value;
    scale = 1 / sqrt(scale);
    //for (int i = 0; i < n; i++)
      //x[i].value *= scale;
  }

  svm_model *model;
  svm_problem problem;
};

class MakeClassifier
{
public:
  MakeClassifier(Classifier &c, bool verbose = false) : c(c), verbose(verbose) {}

  void train(string filelist) {
    map<string, vector<string>> cat2f;
    processFilelist(filelist, [&](string cat, string filename) {
      cat2f[cat].push_back(filename);
    });
    c.train(cat2f);
  }

  void classify(string filelist) {
    int correct = 0, tot = 0;
    processFilelist(filelist, [&](string cat, string filename) {
      string res = c.classify(filename);
      if (verbose)
        cout << filename << ' ' << res << endl;
      if (res == cat) // correct
        correct++;
      tot++;
    });
    cout << "precision: " << double(correct) / tot << endl;
  }

protected:
  void processFilelist(string filelist, function<void(string, string)> cont) {
    ifstream fin(filelist);
    string line, cat;
    while (getline(fin, line)) {
      if (line[0] == '#') {
        cat = line;
        while (! isalnum(cat[0]))
          cat = cat.substr(1);
      } else {
        cont(cat, line);
      }
    }
  }

  Classifier &c;
  bool verbose;
};

int main(int argc, char *argv[])
{
  bool verbose = false;
  if (argc >= 2 && argv[1][0] == '-') {
    verbose = true;
    argc--;
  }
  if (argc == 1) {
    NaiveBayes nb;
    MakeClassifier c(nb, verbose);
    c.train("train.list");
    c.classify("classify.list");
  } else {
    SVM svm;
    MakeClassifier c(svm, verbose);
    c.train("train.list");
    c.classify("classify.list");
  }
}
